zebra-network/src/peer/connection.rs

```diff
@@ -713,110 +752,78 @@ where
                     .send(Message::Inv(hashes.iter().map(|h| (*h).into()).collect()))
                     .await
                 {
-                    Ok(()) => Ok((AwaitingRequest, Some(tx))),
+                    Ok(()) => {
+                        // Since we're not waiting for further messages, we need to
+                        // send a response before dropping tx.
+                        let _ = tx.send(Ok(Response::Nil));
+                        Ok(AwaitingRequest)
+                    },
                     Err(e) => Err((e, tx)),
                 }
             }
             (AwaitingRequest, AdvertiseBlock(hash)) => {
                 match self.peer_tx.send(Message::Inv(vec![hash.into()])).await {
-                    Ok(()) => Ok((AwaitingRequest, Some(tx))),
+                    Ok(()) => {
+                        // Since we're not waiting for further messages, we need to
+                        // send a response before dropping tx.
+                        let _ = tx.send(Ok(Response::Nil));
+                        Ok(AwaitingRequest)
+                    },
                     Err(e) => Err((e, tx)),
                 }
             }
-        };
-        // Updates state or fails. Sends the error on the Sender if it is Some.
-        match new_state_result {
-            Ok((AwaitingRequest, Some(tx))) => {
-                // Since we're not waiting for further messages, we need to
-                // send a response before dropping tx.
-                let _ = tx.send(Ok(Response::Nil));
-                self.state = AwaitingRequest;
-                self.request_timer = Some(sleep(constants::REQUEST_TIMEOUT));
-            }
-            Ok((new_state @ AwaitingResponse { .. }, None)) => {
-                self.state = new_state;
-                self.request_timer = Some(sleep(constants::REQUEST_TIMEOUT));
-            }
-            Err((e, tx)) => {
-                let e = SharedPeerError::from(e);
-                let _ = tx.send(Err(e.clone()));
-                self.fail_with(e);
-            }
-            // unreachable states
-            Ok((Failed, tx)) => unreachable!(
-                "failed client requests must use fail_with(error) to reach a Failed state. tx: {:?}",
-                tx
-            ),
-            Ok((AwaitingRequest, None)) => unreachable!(
-                "successful AwaitingRequest states must send a response on tx, but tx is None",
-            ),
-            Ok((new_state @ AwaitingResponse { .. }, Some(tx))) => unreachable!(
-                "successful AwaitingResponse states must keep tx, but tx is Some: {:?} for: {:?}",
-                tx, new_state,
```

Comment:
The goal of this removed match statement is to make sure that tx and Failed are used correctly.
Where are these checks moved to in this PR?

Reply:
they shouldn't be necessary anymore because I factor'ed out the Option around the `tx`

Reply:
I think I'm going to need a walkthrough and updated doc comments to understand this code.
(Feel free to defer the doc comments until the design has stabilised.)

To be clear, my goal is to make sure that:
1. every request sends a response on `tx`, or we have removed that invariant
2. requests that need responses actually send them, and requests that don't have responses don't send any

I'm not particularly concerned about code duplication.

I created a Request/Response design in #1723 that encodes request timing, response, and routing as types. I'm happy to leave that redesign until later, because it's not causing any bugs right now. But let me know if you'd like me to walk you through it - it might help explain why we treat some requests differently.

Reply:
I think this code is obsoleted, like you originally said.